packages/query/src/block-processor.ts

```diff
@@ -103,89 +115,126 @@ export class BlockProcessor implements BlockProcessorInterface {
+        await this.identifyNewAssets(flush.newNotes);
+
+        // boolean indicates if we should recover source id
+        for (const { noteCommitment, source } of flush.newNotes)
+          blockCommitments.set(
+            uint8ArrayToHex(noteCommitment!.inner),
+            source!.equals(blankTxSource),
+          );
+        for (const { swapCommitment, source } of flush.newSwaps)
+          blockCommitments.set(
+            uint8ArrayToHex(swapCommitment!.inner),
+            source!.equals(blankTxSource),
+          );
       }
 
       // nullifiers on this block may match notes or swaps from db
-      // mark them as spent/claimed, return what was used
-      const spentNullifiers = await this.resolveNullifiers(
+      // - update idb, mark as spent/claimed
+      // - return records' commitments
+      // - with a boolean to indicate if we should recover source id
+      const nullifierCommitments = await this.resolveNullifiers(
         compactBlock.nullifiers,
         compactBlock.height,
       );
-      // commitments on this block?? idk man
-      const commitmentRecords = await this.resolveCommitments(
-        compactScan?.sctUpdates.store_commitments,
-      );
-
-      const fullScanReasons = {
-        'spent nullifiers': spentNullifiers.size,
-        'updated commitments': commitmentRecords.size,
-        'new notes': compactScan?.newNotes.length,
-        'new swaps': compactScan?.newSwaps.length,
-      };
-
-      if (Object.values(fullScanReasons).some(Boolean)) {
-        console.log('full scan', compactBlock.height, fullScanReasons);
 
+      // if anything happend involving a commitment, scan all block tx
+      if (nullifierCommitments.size || blockCommitments.size) {
+        const allCommitments = new Map([...nullifierCommitments, ...blockCommitments]);
+        console.warn(
+          'full block fetch',
+          'all commitments',
+          flush?.sctUpdates.store_commitments.length,
+          'nullifierCommitments',
+          nullifierCommitments.size,
+          'blockCommitments',
+          blockCommitments.size,
+          'allCommitments',
+          allCommitments.size,
+        );
         // this is a network query
-        const fullBlock = await this.querier.app.txsByHeight(compactBlock.height);
+        const { blockHeight, transactions } = await this.querier.app.txsByHeight(
+          compactBlock.height,
+        );
 
-        const relevantTx = await this.identifyRelevantTransactions(
-          Array.from(spentNullifiers.keys()),
-          Array.from(commitmentRecords.keys()),
-          fullBlock.transactions,
+        const { relevantTx, sourceTx } = await this.identifyTransactions(
+          allCommitments,
+          transactions,
         );
 
-        console.warn('relevantTx', relevantTx.length);
+        console.warn('relevantTx', relevantTx.size, 'sourceTx', sourceTx.size);
+
+        await this.saveSourceIds(sourceTx);
```

Comment:
rehydrating source ids is its own function

Reply:
this feels more complex than the combined detect tx / hydrate source as requested in another comment so i will combine